packages/flutter/lib/src/widgets/restoration.dart

```diff
@@ -0,0 +1,1002 @@
+/// [RootRestorationScope] will build an empty [Container] instead of the actual
+/// [child] until the root bucket is available. To hide the empty container from
+/// the eyes of users, the [RootRestorationScope] also delays rendering the
+/// first frame while the container is shown. On most platforms, this keeps the
+/// splash screen up (hiding the empty container) until the bucket is available
+/// and the [child] is ready to be build.
+///
+/// See also:
+///
+///  * [RestorationScope], which inserts a bucket obtained from a surrounding
+///    restoration scope into the widget tree and makes it accessible
+///    for descendants via [RestorationScope.of].
+///  * [UnmanagedRestorationScope], which inserts a provided [RestorationBucket]
+///    into the widget tree and makes it accessible for descendants via
+///    [RestorationScope.of].
+///  * [RestorationMixin], which may be used in [State] objects to manage the
+///    restoration data of a [StatefulWidget] instead of manually interacting
+///    with [RestorationScope]s and [RestorationBucket]s.
+///  * [RestorationManager], which describes the basic concepts of state
+///    restoration in Flutter.
+class RootRestorationScope extends StatefulWidget {
+  /// Creates a [RootRestorationScope].
+  const RootRestorationScope({
+    Key key,
+    @required this.restorationId,
+    this.child,
+  }) : super(key: key);
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  /// The [RestorationId] used to identify the child bucket that this widget
+  /// will insert into the tree.
+  ///
+  /// If this is null, no bucket is made available to descendants and state
+  /// restoration for the subtree is essentially turned off.
+  final RestorationId restorationId;
+
+  @override
+  State<RootRestorationScope> createState() => _RootRestorationScopeState();
+}
+
+class _RootRestorationScopeState extends State<RootRestorationScope> {
+  bool _okToRenderBlankContainer;
+  RestorationBucket _rootBucket;
+  RestorationBucket _ancestorBucket;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _ancestorBucket = RestorationScope.of(context);
+    _loadRootBucketIfNecessary();
+    _okToRenderBlankContainer ??= widget.restorationId != null && _needsRootBucketInserted;
+  }
+
+  @override
+  void didUpdateWidget(RootRestorationScope oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _loadRootBucketIfNecessary();
+  }
+
+  bool get _needsRootBucketInserted => _ancestorBucket == null;
+
+  bool get _isWaitingForRootBucket {
+    return widget.restorationId != null && _needsRootBucketInserted && _rootBucket == null;
+  }
+
+  bool _isLoadingRootBucket = false;
+
+  void _loadRootBucketIfNecessary() {
+    if (_isWaitingForRootBucket && !_isLoadingRootBucket) {
+      _isLoadingRootBucket = true;
+      RendererBinding.instance.deferFirstFrame();
```

Comment:
What happens if this is inserted into the tree after frames have already started rendering?

Reply:
Also: what happens if anything below this throws? Should we try to resume frames, or should we just have a hard crash? Right now it seems like an app could get into a state where it throws in here and just never renders frames again.